pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -595,6 +837,9 @@ func (pl *DynamicResources) Filter(ctx context.Context, cs fwk.CycleState, pod *
 
 	if state.allocator != nil {
 		state.nodeAllocations[node.Name] = allocations
+		if pl.enableExtendedResource {
+			state.nodeExtendedResourceClaim[node.Name] = extendedResourceClaim
```

Comment:
The code above checks whether it needs to update something before locking the mutex. You need to extend that if check, otherwise the mutex might not be locked when you get here. Also, I think you can replace the feature gate check with a check for `extendedResourceClaim` being nil. If the feature is disabled, it will be nil. If it's enabled, we don't need to store a nil pointer (which might not occur, but it wouldn't make sense if it did).

Reply:
That's inside an `if state.allocator != nil || len(unavailableClaims) > 0 {` check.

If there are no unavailable claims, the mutex will not be locked.

Reply:
Never mind, I see what you mean. That's an `||`, not an `&&`.

We should clean up the `state.allocator != nil` checks, but not in this PR. It's always non-nil nowadays.